src/components/FeatureStyler/FeatureStyler.js

```diff
@@ -277,61 +277,77 @@ class FeatureStyler extends PureComponent {
-
-      if (textSize) {
-        textStyle.setScale(textSize.scale);
+    // Return a promise in case the image needs to be loaded.
+    return new Promise(resolve => {
+      // Update Fill style if it existed.
+      const fillStyle = oldStyle.getFill();
+      if (fillStyle && color) {
+        fillStyle.setColor(color.fill.concat(fillStyle.getColor()[3]));
       }
 
-      if (font) {
-        textStyle.setFont(font);
+      // Update Stroke style if it existed
+      const strokeStyle = oldStyle.getStroke();
+      if (strokeStyle && color) {
+        strokeStyle.setColor(color.fill.concat(strokeStyle.getColor()[3]));
       }
 
-      if (textColor) {
-        const olColor = textColor.fill.concat([1]);
-        const textFill = textStyle.getFill();
-        textFill.setColor(olColor);
-        textStyle.setFill(textFill);
-      }
+      // Update Text style if it existed;
+      const textStyle = oldStyle.getText() ? oldStyle.getText().clone() : null;
+      if (textStyle) {
+        textStyle.setText(text);
 
-      textStyle.setRotation(textRotation);
-    }
+        if (textSize) {
+          textStyle.setScale(textSize.scale);
+        }
 
-    // Update Icon style if it existed.
-    let iconStyle = oldStyle.getImage();
-    if (iconStyle instanceof Icon && icon) {
-      iconStyle = new Icon({
-        src: icon.url,
-        scale: iconSize.scale,
-        anchor: icon.anchor,
-      });
-
-      // We load the icon manually to be sure the size of the image's size is set asap.
-      // Useful when you use a layer's styleFunction that makes some canvas operations.
-      iconStyle.load();
-    }
+        if (font) {
+          textStyle.setFont(font);
+        }
 
-    return new Style({
-      fill: fillStyle,
-      stroke: strokeStyle,
-      text: textStyle,
-      image: iconStyle,
-      zIndex: oldStyle.getZIndex(),
+        if (textColor) {
+          const olColor = textColor.fill.concat([1]);
+          const textFill = textStyle.getFill();
+          textFill.setColor(olColor);
+          textStyle.setFill(textFill);
+        }
+
+        textStyle.setRotation(textRotation);
+      }
+
+      // Update Icon style if it existed.
+      let iconStyle = oldStyle.getImage();
+      if (iconStyle instanceof Icon && icon) {
+        iconStyle = new Icon({
+          src: icon.url,
+          scale: iconSize.scale,
+          anchor: icon.anchor,
+        });
+
+        if (!iconStyle.getSize()) {
+          // Ensure the image is loaded before applying the style.
+          iconStyle.load();
+          return iconStyle.getImage().addEventListener('load', () => {
+            resolve(
+              new Style({
+                fill: fillStyle,
+                stroke: strokeStyle,
+                text: textStyle,
+                image: iconStyle,
+                zIndex: oldStyle.getZIndex(),
+              }),
+            );
+          });
+        }
+      }
+      return resolve(
+        new Style({
```

Comment:
I guess, you can create a style variable before  the  previous if and use it to avoid duplication of Style declaration